src/main/scala/io/iohk/ethereum/db/storage/ReferenceCountNodeStorage.scala

```diff
@@ -0,0 +1,132 @@
+package io.iohk.ethereum.db.storage
+
+import akka.util.ByteString
+import io.iohk.ethereum.db.storage.NodeStorage.{NodeEncoded, NodeHash}
+import io.iohk.ethereum.mpt.NodesKeyValueStorage
+import io.iohk.ethereum.rlp.RLPImplicitConversions.{toRlpList, _}
+import io.iohk.ethereum.rlp.RLPImplicits._
+import io.iohk.ethereum.rlp.{encode => rlpEncode, _}
+
+/**
+  * This class is used to store Nodes (defined in mpt/Node.scala), by using:
+  * Key: hash of the RLP encoded node
+  * Value: the RLP encoded node
+  */
+class ReferenceCountNodeStorage(nodeStorage: NodeStorage, blockNumber: Option[BigInt] = None) extends NodesKeyValueStorage {
+
+  import ReferenceCountNodeStorage._
+
+  override def get(key: ByteString): Option[NodeEncoded] = nodeStorage.get(key).map(_.toStoredNode).map(_.nodeEncoded.toArray)
+
+  override def update(toRemove: Seq[NodeHash], toUpsert: Seq[(NodeHash, NodeEncoded)]): NodesKeyValueStorage = {
+
+    require(blockNumber.isDefined)
+
+    val bn = blockNumber.get
```

Comment:
I was expecting this coment :smile: 

- If you are using Basic pruning mode, it will always store alongside with the mpt node, the reference count. That means that we are "changing" what data is being stored by MPT and it's the reason why you cant switch between archive and basic without reseting the database
- Because of :arrow_up: , everytime you want to access to the mpt nodes, you need to use the corresponding `NodesKeyValueStorage` or it won't work.
- Sometimes, when we need to access the state just to read some value (for example: `io.iohk.ethereum.domain.BlockchainImpl#getAccountStorageAt`) we don't have (and don't need) to tag the changes under a certain block number. 
- Other alternatives were considered but I wanted to avoid adding block taging changes within MPT (because it doesn't care if you are using a pruning mechanism or whatever) and that's why it's on `ReferenceCountNodeStorage` constructor and not just in update (which would be more logical).

Sorry for the long explanation, let me know if you need some clarification. Any other ideas are welcome :smile: